style-guide.md

```diff
@@ -0,0 +1,65 @@
+    - It is *not* allowed to have a resource which can be allocated *either* statically or dynamically.
+      - This is technically feasible, but is too complicated for maintenance and end-users.
+      - e.g. `memory_gb` and `modify_memory_gb` cannot be present in the same task.
+- All tasks and workflows should have a `max_retries` input.
+  - TODO: again, only enforeceable with specific var names. Not sure if we want to get into that business. Keep it in the guide?
+  - This should be defaulted to `1` for nearly all tasks
+  - Some tasks are particularly error prone and can have a higher default `max_retries`
+  - **rule specific to workflows:** `max_retries` should be an optional `Int?`
+    - This allows each task to have it's own specific default `max_retries`
+      - If a user does not supply `max_retries`, those task level defaults will get used
+      - If a user does supply `max_retries`, it should override the default for *every* task called
+- multi-core tasks should *always* follow the conventions laid out in the `detect_nproc_task` example (see `template/task-templates.wdl`)
+  - this is catering to cloud users, who may be allocated a machine with more cores than are specified by the `ncpu` parameter
+  - TODO: this would be tough to validate with a linter. It is definitely a best practice, but IMO doesn't belong in the guide
+- `command` blocks should be wrapped with arrows (`<<< >>>`) instead of brackets (`{ }`)
+  - Certain Bash constructions cause problems with the bracket notation
+- output file names should *always* be determined with either the `outfile_name` parameter or the `prefix` parameter.
+  - `outfile_name` should be preferred if no downstream tasks/tools rely on the file name/extension
+  - tasks with multiple outputs should always use the `prefix` convention
+  - TODO: IMO another "best practice" that shouldn't be enforced by a linter. Remove from guide?
+- output variable names should be short but descriptive
+  - TODO: This I'm more inclined to enforce with the linter. I think disallowing specific output names is better than forcing users into using specific ones.
+  - Disallowed output names: `out*`, any name less than 3 characters
+    - TODO: is this sufficient?
+- All tasks should run in a Docker container
+  - TODO: is there a valid reason not to use a Docker container? Should this be enforced by the linter?
+- no whitespace on empty lines
+- no whitespace at the end of lines
+- WDL lines should be less than 100 characters wide whenever possible
+  - Exceptions would be long strings that WDL doesn't allow to be broken up
+  - This restriction applies to embedded code in the `command` block as well.
+    - TODO: I'm still not sure about this decision. I have reservations applying this rule to the command block. I've found a small handful of cases where I think the most readable formatting extends over 100 characters. They are rare, but exist frequently enough I'm not sure I like this rule.
```

Comment:
`parameter_meta` is also likely to exceed this mark, for certain parameters. For example below, we have
```
        modify_disk_size_gb: "Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB."
```

Reply:
Yes, `parameter_meta` will have to be excluded from this check. WDL v1 does not offer a way to insert line breaks in long strings for this section. There will probably be other exceptions as well, though I can't think of any atm. In these cases where conforming to the WDL spec involves breaking the 100 character limit, we'll just need to ignore it. Nothing we can do that doesn't break things.

_But_, the `command` block can theoretically ALWAYS be forced into under 100 characters. I just don't think that that always leads to the most readable code. It does _most_ of the time, but there are exceptions. For example, we have a few lines like the below:
```
        kraken2-build --build \
            ~{if protein then "--protein" else ""} \
            --kmer-len ~{kmer_len} \
            --minimizer-len ~{minimizer_len} \
            --minimizer-spaces ~{minimizer_spaces} \
            ~{if (max_db_size_gb > 0) then "--max-db-size" else ""} ~{if (max_db_size_gb > 0) then max_db_size_bytes else ""} \
            --threads "$n_cores" \
            --db ~{db_name}
```
IMO this is the most readable format we could write, and it's 128 characters. It could be rewritten to be under 100 like so:
```
            --minimizer-spaces ~{minimizer_spaces} \
            ~{if (max_db_size_gb > 0) then "--max-db-size" else ""} \
                ~{if (max_db_size_gb > 0) then max_db_size_bytes else ""} \
            --threads "$n_cores" \
```
But IMO that's harder to follow.

Maybe we could add a way to the linter to disable the error on a per-line basis? Ending the line with something like `# --ignore-long-line`? Problem with this is it compounds the problem. We'd be going "this line is readable when it's long so it should be an exception. To signify that we'll make it _EVEN LONGER_" 😆 

Maybe we have to accept that _rarely_, forcing a line under 100 characters will be slightly worse, but it leads to more readable code the majority of the time.

We have some tasks that were written like:
```
tool_foo --parameter1 {foo} --parameter2 {foo} --parameter3 {foo} etc.
```
that extend well past 100 characters and are just terrible to look at. I most definitely want our tool to flag those lines as needing to be rewritten.

Not sure what the answer is here. I just threw all my thoughts up against the wall here. Do you have any thoughts to contribute on this?

Reply:
Ok, I spent some time weighing the options here, and I've decided that I'm ok with enforcing character limits within the `command` block. Yes, there will be a handful of rare cases where this makes things less readable, but the collapsed alternatives aren't worse enough to justify dealing with making exceptions.

However I was reading through the the [`black` rules](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#), mainly looking for whitespace rules to steal (and those have been committed), but they have a section justifying why their character limit is 88 wide. I chose 100 pretty arbitrarily. `black` allows the line length to be configured, not sure if we want to deal with that or just pick something sane and enforce it. But here's what they have to say about lengths longer than the 88 default:

> You can also increase it, but remember that people with sight disabilities find it harder to work with line lengths exceeding 100 characters. It also adversely affects side-by-side diff review on typical screen resolutions. Long lines also make it harder to present code neatly in documentation or talk slides.

Now I'm considering dropping our line length from 100 to 90. "88" is just a weird number to choose if you ask me, so my opinion is to take their suggestion and round to a nicer number. That's 90.

@adthrasher any opinion on either point I make here?

Reply:
Thanks for looking that up. The two points made by `black` are good. I'm fine with picking a round number. Their documentation also says "90-ish" is a good choice, so using the round number seems fine.